### ISSUE TITLE
WirePattern - Should this be a base component?

### ISSUE THREAD

**Issue by hayesmg** [noflo/noflo#205]
It seems that the WirePattern functionality does quite a bit in terms of dealing with input data.  Is it possible to get a base component that offers this functionality?  That base WirePattern component could then be configured by inspecting metadata and/or ports.

For example it would be nice to identify a port as "required for output" or some other nomenclature that signifies arrival of its data must happen first.  Here is an example that borrows from the example code in CHANGES.md for 0.5.6.  My apologies for inexperience with coffee-script.

```coffee
component = new noflo.WirePatternComponent
component.inPorts.add 'path',
  datatype: 'string'
  required: true
  requiredForOutput: true
component.inPorts.add 'delay',
  datatype: 'int'
  required: false
  requiredForOutput: true
component.inPorts.add 'line',
  datatype: 'string'
component.inPorts.add 'repeat',
  datatype: 'int'
component.outPorts.add 'out',
  datatype: 'object'
component.outPorts.add 'error',
  datatype: 'object'

component.config.async = true

component.config.wirePatternCallback = (data, groups, out, callback) ->
  # This function is the callback for the WirePattern functionality.
```

**Comment by jonnor** [comment]
@trustmaster created some sketches/proposals for new Component APIs here: https://gist.github.com/trustmaster/2a292e23e61642fa148d

**Comment by jonnor** [comment]
Some quick feedback/idea first. I see that Promises are used in the proposals, which I think is good. I'd be fine with requiring Promises for async stuff (can use `promisify` function if one doesn't have one already).

But why don't we let the NoFlo Component/HelperFoo handle .catch and calling done() - by attaching to the Promise? Less verbose and error-prone, though routing to many outports or doing special things with error becomes hard(er) - is that really important?

**Comment by jonnor** [comment]
Another quick one. I think the most classical FBP style component would be to have `InPort.read()` be a function which takes port name, and return a promise which will resolve when data comes on that inport.

**Comment by trustmaster** [comment]
@jonnor Promises are not used in the proposed API itself, it's `doSomething()` example function which is invoked inside of process body that returns a promise. It doesn't affect the API itself, it's just a mere example of async code which could be callback-style or synchronous as well.

I'm a bit hesitant of using Promises in NoFlo itself, at least for sake of older browser compatibility, extra dependencies and performance/memory footprint.

**Comment by jonnor** [comment]
@trustmaster OK, fair enough I think. Only the super-classical FBP style is ruled out by such a choice. (...writing more tricky feedback)

**Comment by jonnor** [comment]
For more context, we also have some earlier proposals/discussions here:
https://docs.google.com/document/d/1_7ExwbugP8Q8LNaaysRwWxny9sgb9FomXL0Vs8sU0cc/edit#heading=h.ggqvo151ftpf

**Comment by trustmaster** [comment]
TLDR from jpaulm/jsfbp#14 and jpaulm/jsfbp#25: blocking reads can be emulated in plain JavaScript using Promises relatively easily, but back pressure is rather complicated. It would be nice to have super-classical option, but not worth sacrificing simplicity and performance IMO.

**Comment by jonnor** [comment]
Now for some more tricky/general discussion. I think there are several interface/API-pieces involved here. Primary being: 1) Getting data from input queues 2) When/how to process the collected input data (firing pattern)*.
The first has implications only for inside the component, where as the second also affects the outside world - because the semantics of the component changes. And thus influencing a lot how one programs with NoFlo on the graph level. So I think that is the most critical to get right...

For instance, one thing I think is bad right now, is that it is the Component - instead of the graph using it - that decides which ports are Controls (non-triggering) and which are not. I fear that trying to have a general / very configurable firing pattern mechanism cause such problems to explode.
So, can we try to define what kind of firing patterns we think are important to support, and which ones are too esoteric to have first-level support for (the rare cases can use old basic/manual API)?

I give it a go here. Important:
* `map` - takes input on 1 port, transform into something new (or sends same on)
* `combineLast` - takes 1 input on each of the N ports, creates something new
* `combineSameRequest` - takes 1 input on each of the N ports, *matched by IP correlation ID*, creates something new
* more? maybe some sort of splitting?

Not important:
* Sending errors on arbitrary outports. I think all failures, should go a port name/type established by convention, like `error`.
* Matching on groups instead of correlation id. If wanting this, use a dedicated component, which changes the correlation ID to be that of group[N].


`*` the last part of Component API, declaring information about component for outside world (ports etc) is not up for discussion now/here. As I see it, it mostly works fine.

**Comment by trustmaster** [comment]
I would argue that it is a graph author who decides when a component should react to its input.

I often use components that have strong separation of processed input and controls. For example, a component that writes a post on behalf of a user into database. Inputs: `post`, `user` objects. Controls: `database` connection handle, and some static data like `tableName`. A component decides when to write what and where. If it was up to a graph, there could be a situation when it would make a meaningless write on database connection reset, for instance.

In IDEF0, difference between input and controls is visible on graph level (input arrows enter the box on the left, controls enter it on the top side), but input and controls can't be swapped because they are defined in the component. In Classical FBP, all the firing logic and component behavior is encapsulated in process function and the only way to tell a graph author about port specifics is by documenting it. CFBP doesn't lean towards super-reusable microcomponents, or super-configurable macrocomponents, so a graph author deducts what a component does and what input it expects either from documentation, or from code, or by a mere assumption if it's simple enough.

**Comment by jonnor** [comment]
Who is the component developer to say that `database` must always be a `Control`. What if I (the graph developer, and user of the component) want to shard database access, say based on user groups. Then I need to specify a which database to use per request?

**Comment by trustmaster** [comment]
Swapping controls to inputs and vice versa in the graph would sometimes make sense, yes. But a component author should be very careful about it then, providing edge cases for every possibility. E.g. database components must always be written with data property based sharding in mind. Aren't we introducing too much complexity here?

**Comment by jonnor** [comment]
Database components would just take a DB instance (or connection handle), and call some method on it. I don't see why it would require special-casing.

With current-style Control handling, I do not see how we can provide facilities to treat them as inputs.
We don't have a way to let data on control use matching on correlation ids, like inputs have?

**Comment by trustmaster** [comment]
All IPs are now objects which may have `scope`. Control IPs is not an exception.

**Comment by jskrzypek** [comment]
@jonnor, @trustmaster, 
Regarding the idea to add `InPort.read([portname]) = () => Promise<input>`:

You may have already considered and dismissed this idea, but here goes:

I assume you guys are familiar with Kris Kowal's GTOR (https://github.com/kriskowal/gtor) and related reactive javascript libraries, [RxJS](https://github.com/Reactive-Extensions/RxJS) or [Kefir.js](https://rpominov.github.io/kefir/)? 

If you're open to rebuilding some of the internals of noflo (I think mostly just InternalSocket, though), I think the much better upgrade would be to have OutPorts embed `Observable`s that `subscribe()` attached InPorts.

The beautiful thing about doing something like this on top of RxJS is that you'd get a number of backpressure options out-of-the box, see their [backpressure.md](https://github.com/Reactive-Extensions/RxJS/blob/master/doc/gettingstarted/backpressure.md).

In terms of browser capability, RxJS is a component one of the major dataflows that people are using in Angular 2 apps, and for IoT compatability, there are a number of "lite" RxJS libraries you can use to limit the footprint.

**Comment by trustmaster** [comment]
@jskrzypek you're right, I've already dismissed Promises as current priority because without backpressure they have almost no benefits. But now I have something to study over the weekend, thanks for sharing this!

Do you have any idea of what performance penalty it could be to base `InternalSocket` on RxJS?

**Comment by jskrzypek** [comment]
Also, regarding your discussion of input vs control:

I have an idea for a nice pattern that might make sense here and satisfy both sides of the discussion. It kind of depends on and follows from making the Ports/InternalSockets into Observables, but I think there are ways to implement it even if you reject that Proposal.

At a high level, `component`s that want to use Control ports should implement a `configureProcess()` method instead of a `process()` method. `configureProcess()` would be triggered by the Control ports and set the component's `process()` function. Building on the scenario from @trustmaster 's gist above, here's a sketch of what it would do.

```js
## sync configureProcess
c.configureProcess = (controls) ->
  doConfig controls
  (input, output, done) -> 
    doSomething input.user.data, input.post.data, timeout: control.timeout
    .then (result) ->
      output.result.data result,
        groups: input.post.groups
      done()
    .catch done
## async configureProcess
c.configureProcess = (controls, configDone) ->
  doConfigAsync controls, ->
    (input, output, done) -> 
      doSomething input.user.data, input.post.data, timeout: control.timeout
      .then (result) ->
        output.result.data result,
          groups: input.post.groups
        done()
      .catch done
```

If you agree with the idea of building the ports around Observables, this way of doing things becomes even cleaner, since in the reactive scenario, `configureProcess()` is subscribed to the Control ports' `Observables`, and itself returns an `Observable`. Then the component subscribes to the Input ports' `Observable`s with the `process()` function that it sees coming from `configureProcess()` as the `Observer` to the inputs.

**Comment by trustmaster** [comment]
I've settled with a Controls idea as simple as this:

- `control` flag changes behavior of `InPort.buffer` and  `InPort.receive()`: when it's `false` (default) the buffer stores all values and `receive()` shifts the packet off the buffer; when it's `true` the buffer stores the current value and `receive()` doesn't remove the packet from buffer. The flag can be flippable on Graph/Flowhub level
- `triggering` flag controls another aspect: when `true` (default) a new packet on the inport does trigger the process function, when `false` it doesn't

**Comment by jskrzypek** [comment]
I don't know a lot about the performance implications of RxJS, but I did find some discussion about their performance available here: https://github.com/Reactive-Extensions/RxJS/blob/master/doc/mapping/bacon.js/whyrx.md

**Comment by jonnor** [comment]
Using RxJS on that level, where inports subscribe directly to outports, would take us even further away from the bounded-queues concept of FBP. I really think we should move towards having queues more observable, not less.

**Comment by jskrzypek** [comment]
You can have the RxJS/Observables and still use the bounded-queues, you just use the InternalSockets as intermediaries with queues.

Here's a pattern I found that's basically doing that (albeit from the java world).

http://www.javaworld.com/article/2077526/build-ci-sdlc/java-tip-29--how-to-decouple-the-observer-observable-object-model.html

**Comment by trustmaster** [comment]
Bounded Queues + Observables, hmm, isn't this getting rather complicated? Time to slow down and ask a question: what problem are we trying to solve?

Most likely it is getting 2 missing characteristics of Classical FBP in NoFlo: blocking reads and backpressure (which is blocking writes in CFBP). Let's see what these 2 are used for in CFBP:

- _Blocking reads_ suspend the downstream process until the data from exact inport arrives. Given the multithreaded long-running nature of CFBP, this is the way to program firing patterns within the process function itself. Given single-threaded EventLoop in JavaScript, this probably isn't the desired way to program firing patterns in JavaScript. In JS non-loopers are more simple and lightweight than loopers, so checking for a pattern match on every packet and calling the process function is what NoFlo currently leans toward: why use Promises and Observables for the job that the Event Loop does for free.
- _Blocking writes_ suspend the upstream process when the downstream process is busy and the connection buffer is full. This is just one of many potential solutions to rate limiting and whether it is lossless or lossy depends on the upstream process details. E.g. if the upstream reads from file, it can safely idle when paused, but if it listens to UI events, some of those events will be definitely lost when paused. As you see, the problem of rate limiting is application specific and _blocking writes_ is not a silver bullet for it. _RxJS_ actually provides more options for backpressure than CFBP does, but it goes far beyond the simple concept of a bounded buffer and ports on its ends. Keeping rate limiting simple and yet effective is an open question for now.

**Comment by gabejohnson** [comment]
@trustmaster you may want to check out

https://github.com/ubolonton/js-csp

It's a Clojure core.async inspired library. It uses Go style channels for managing concurrency.

https://en.wikipedia.org/wiki/Go_(programming_language)#Concurrency:_goroutines_and_channels

I've used core.async in Clojure and ClojureScript and find it very useful for message buffering and sequencing.

**Comment by trustmaster** [comment]
@gabejohnson js-csp looks interesting, definitely worth trying and considering for e.g. NoFlo 2.0.

**Comment by bergie** [comment]
OK, we've settled on Process API (#392) now.
